src/shogun/features/DenseFeatures.h

```diff
@@ -134,6 +137,27 @@ template<class ST> class CDenseFeatures: public CDotFeatures
 	 */
 	SGVector<ST> get_feature_vector(int32_t num) const;
 
+	/**
+	 * Returns an iterator to the first vector of features.
+	 */
+	iterator begin() noexcept
+	{
+		return SubsetIterator(this);
+	}
+
+	/**
+	 * Returns an iterator to the element following the end of features.
+	 */
+	iterator end() noexcept
+	{
+		if (auto stack = get_subset_stack()->get_last_subset();
+			stack == nullptr)
+			return SubsetIterator(this, get_num_vectors());
```

Comment:
any chance this can be shortened/moved inside the iterator?

Reply:
yup, put this in the crtp base class